virtualmachine/vsphere/vm.go

```diff
@@ -1598,90 +1661,114 @@ func GetTemplateList(vm *VM) ([]map[string]interface{}, error) {
+			}
+			devinfo, ok := disk.DeviceInfo.(*types.Description)
+			if !ok {
+				continue
 			}
+			backing := disk.Backing
+			fileBackingInfo := backing.(types.BaseVirtualDeviceFileBackingInfo).GetVirtualDeviceFileBackingInfo()
+			diskInfo = append(diskInfo, map[string]interface{}{
+				"name":      devinfo.Label,
+				"size":      disk.CapacityInKB,
+				"disk_file": fileBackingInfo.FileName,
+			})
 		}
+		vmList = append(vmList, map[string]interface{}{
+			"name":           name, // full name/path of vm
+			"id":             vmo.Self.Value,
+			"disks":          diskInfo,
+			"nic_info":       getNicInfo(vmo),
+			"memory_size_mb": vmo.Summary.Config.MemorySizeMB,
+			"num_cpu":        vmo.Summary.Config.NumCpu,
+		})
 	}
 	return vmList, nil
 }
 
-// getVirtualMachines : Return the virtual machines in a cluster
-func getVirtualMachines(vm *VM) ([]mo.VirtualMachine, error) {
+// getVirtualMachines : Returns the virtual machines in a dc/cluster/host
+func getVirtualMachines(vm *VM) (map[string]mo.VirtualMachine, error) {
 	var (
-		vmList          []mo.VirtualMachine
-		virtualMachines []mo.VirtualMachine
-		hsMo            mo.HostSystem
+		vmsInDc, vmsInCluster map[string]mo.VirtualMachine
+		hsMos                 []mo.HostSystem
+		hsMo                  mo.HostSystem
 	)
+	vmsInDc = make(map[string]mo.VirtualMachine)
+	vmsInCluster = make(map[string]mo.VirtualMachine)
 	// set up session to vcenter server
 	if err := SetupSession(vm); err != nil {
 		return nil, err
 	}
 
-	if vm.Destination.DestinationName == "" {
-		// Return virtual machines for the whole datacenter
-		dcMo, err := GetDatacenter(vm)
-		if err != nil {
-			return nil, err
-		}
-
-		dcObj := object.NewDatacenter(vm.client.Client, dcMo.Reference())
-		return getDcVMList(vm, dcObj)
+	// Return virtual machines for the whole datacenter
+	dcMo, err := GetDatacenter(vm)
+	if err != nil {
+		return nil, err
 	}
 
-	// set up session to vcenter server
-	dc, err := GetDatacenter(vm)
+	dcObj := object.NewDatacenter(vm.client.Client,
+		dcMo.Reference())
+	vmsInDc, err = getDcVMList(vm, dcObj)
 	if err != nil {
 		return nil, err
 	}
 
-	// Get the cluster resource and its host, datastore and datastore
-	crMo, err := findClusterComputeResource(vm, dc, vm.Destination.DestinationName)
+	// if cluster name not given, return vms in datacenter
+	if vm.Destination.DestinationName == "" {
+		return vmsInDc, nil
+	}
+
+	// Get the cluster resource and its host, datastore
+	crMo, err := findClusterComputeResource(vm, dcMo,
+		vm.Destination.DestinationName)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, host := range crMo.Host {
-		err = vm.collector.RetrieveOne(vm.ctx, host, []string{"name", "vm"}, &hsMo)
-		if err != nil {
-			return nil, err
+	// get the hosts in cluster
+	err = vm.collector.Retrieve(vm.ctx, crMo.Host, []string{"name"}, &hsMos)
+	if err != nil {
+		return nil, err
+	}
+	hosts := make(map[string]string)
+	for _, host := range hsMos {
+		hosts[host.Name] = ""
```

Comment:
[NTH] This looks like a hacky way. Is there an alternate approach?
Or add code comment about why this is required.

Reply:
the map here is used for faster/efficient lookup of the hostname. We could use an array of strings and lookup for strings in the array but that will be less efficient than the maps. Will update the comment here.